packages/backend/migrations/20240731084359_add_unique_constraint_to_kid.js

```diff
@@ -0,0 +1,29 @@
+/**
+ * @param { import("knex").Knex } knex
+ * @returns { Promise<void> }
+ */
+exports.up = function (knex) {
+  // delete any existing duplicates per wallet address
+  knex.raw(`
+  DELETE FROM walletAddressKeys
+    WHERE ctid NOT IN (
+      SELECT MIN(ctid)
+      FROM walletAddressKeys
+      GROUP BY walletAddressId, kid, x
+    );
+  `)
+
+  return knex.schema.alterTable('walletAddressKeys', (table) => {
+    table.unique(['walletAddressId', 'kid', 'x'])
+  })
+}
```

Comment:
noticed this wasn't actually running the DELETE (and when it does we need to quote the camelCase names. Need to do something like:


```suggestion
exports.up = function (knex) {
  // delete any existing duplicates per wallet address
  return knex
    .raw(
      `
    DELETE FROM "walletAddressKeys"
      WHERE ctid NOT IN (
        SELECT MIN(ctid)
        FROM "walletAddressKeys"
        GROUP BY "walletAddressId", kid, x
      );
    `
    )
    .then(() => {
      return knex.schema.alterTable('walletAddressKeys', (table) => {
        table.unique(['walletAddressId', 'kid', 'x'])
      })
    })
}
```

Its kind of annoying but when I want to test migrating data I:
1. move the file out of the migration dir
2. spin up the containers and add whatever data that needs to be migrated (adding a duplicate key in this case)
3. stop the containers (dont remove the volume) and move migration file back in
4. spin back up and check

Reply:
Yes, in PostgreSQL you need to put camel case in quotes. Delete statement was not performed probably because of the missing await, because it is async operation.

Reply:
wish we would have used the snake case mapper when we set the db initially: https://vincit.github.io/objection.js/recipes/snake-case-to-camel-case-conversion.html

Reply:
I'll add the quotes.

@golobitch all of these have to be async functions?

Reply:
With knex and objection, if you dont await the function the query is not executed. Instead it returns a query builder (so you can modify further). 

That's what happened originally and asserting as much here really made me stop and think how that was even possible (aka, "am I wrong???"). Guess I simply was not familiar with `then()` ([MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/then)).

- A SO post asking the same thing: https://stackoverflow.com/questions/65319822/how-does-knex-await-execute-a-database-query
- a little knex query builder prototype that connects all the dots

Reply:
I've found there to be a lot of gotchas with promises in the knex migrations. [Another important thing is returning from the migration](https://github.com/knex/knex/issues/3165#issuecomment-486646428) (which we are doing). We couldn't, for example, await all the queries and not return (something I've done).

I've also seen/tried using `Promise.All` like this but I don't think it would guarantee resolution order because they run concurrently (ie, second might actually modify the db before the first):

```ts
return Promise.All([
  knex.raw(...),
  knex.raw(...)
])
```

Usually I actually use the dreaded promise chaining:

```ts
return knex.raw(...)
 .then(()=>{
   knex.raw(...)
 })
```

Ugly, I know. But clear.

The way it is now seems fine - just wanted to chime in with other general pitfalls I've run into.

Reply:
@sabineschaller No, I meant that `knex.raw` needs to be awaited